rtgui/batchqueue.cc

```diff
@@ -509,107 +400,106 @@ int cancelItemUI (void* data)
+        for (const auto item : items) {
+
+            const auto entry = static_cast<BatchQueueEntry*> (item);
 
-            if (entry->processing) {
+            if (entry->processing)
                 continue;
-            }
 
-            std::vector<ThumbBrowserEntryBase*>::iterator pos = std::find (fd.begin(), fd.end(), entry);
+            const auto pos = std::find (fd.begin (), fd.end (), entry);
 
-            if (pos != fd.end()) {
-                fd.erase (pos);
-                rtengine::ProcessingJob::destroy (entry->job);
+            if (pos == fd.end ())
+                continue;
 
-                if (entry->thumbnail) {
-                    entry->thumbnail->imageRemovedFromQueue ();
-                }
+            fd.erase (pos);
 
-                g_idle_add (cancelItemUI, entry);
-            }
-        }
+            rtengine::ProcessingJob::destroy (entry->job);
 
-        for (size_t i = 0; i < fd.size(); i++) {
-            fd[i]->selected = false;
+            if (entry->thumbnail)
+                entry->thumbnail->imageRemovedFromQueue ();
+
+            g_idle_add (cancelItemUI, entry);
         }
 
-        lastClicked = NULL;
+        for (const auto entry : fd)
+            entry->selected = false;
+
+        lastClicked = nullptr;
         selected.clear ();
     }
 
-    saveBatchQueue( );
+    saveBatchQueue ();
 
     redraw ();
     notifyListener (false);
 }
 
-void BatchQueue::headItems (std::vector<ThumbBrowserEntryBase*>* items)
+void BatchQueue::headItems (const std::vector<ThumbBrowserEntryBase*>& items)
 {
     {
-        // TODO: Check for Linux
 #if PROTECT_VECTORS
         MYWRITERLOCK(l, entryRW);
 #endif
+        for (auto item = items.rbegin(); item != items.rend(); ++item) {
 
-        for (int i = items->size() - 1; i >= 0; i--) {
-            BatchQueueEntry* entry = (BatchQueueEntry*)(*items)[i];
+            const auto entry = static_cast<BatchQueueEntry*> (*item);
 
-            if (entry->processing) {
+            if (entry->processing)
                 continue;
-            }
 
-            std::vector<ThumbBrowserEntryBase*>::iterator pos = std::find (fd.begin(), fd.end(), entry);
+            const auto pos = std::find (fd.begin (), fd.end (), entry);
 
-            if (pos != fd.end() && pos != fd.begin()) {
-                fd.erase (pos);
+            if (pos == fd.end () || pos == fd.begin ())
+                continue;
 
-                // find the first item that is not under processing
-                for (pos = fd.begin(); pos != fd.end(); pos++)
-                    if (!(*pos)->processing) {
-                        fd.insert (pos, entry);
-                        break;
-                    }
-            }
+            fd.erase (pos);
+
+            // find the first item that is not under processing
+            const auto newPos = std::find_if (fd.begin (), fd.end (), [] (const ThumbBrowserEntryBase* fdEntry) { return !fdEntry->processing; });
```

Comment:
Nice one too, using Lambda function.

Wouldn't it be interesting to add a special "//SAMPLE: brief explanation" comment above the key elements that you want to show so that we could make a global search for //SAMPLE and pick in the result list the piece of code that will make our life easier ? I'd add one above this line.